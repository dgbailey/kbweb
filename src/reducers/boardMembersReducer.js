import {
	ADD_BOARD_MEMBER_START,
	ADD_BOARD_MEMBER_SUCCESS,
	ADD_BOARD_MEMBER_FAILURE
} from '../actions/addMemberToBoard';

const boardMemberState = {
	addMemberStart: false,
	addMemberSuccess: false,
	addMemberFailure: false
};
export const boardMembersReducer = (boardMemberState, action) => {
	switch (action.type) {
		case 'ADD_BOARD_MEMBER_START':
			return {
				...state,
				addMemberStart: true,
				addMemberStart: false,
				addMemberFailure: false
			};
		case 'ADD_BOARD_MEMBER_SUCCESS':
			return {
				...state,
				addMemberSuccess: true,
				addMemberStart: false
			};
		case 'ADD_BOARD_MEMBER_FAILURE':
			return {
				...state,
				addMemberSuccess: true,
				addMemberStart: false
			};
		default:
			return state;
	}
};
