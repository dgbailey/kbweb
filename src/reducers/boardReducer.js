import uuid4 from "uuid4";
import {
  ADD_COL_START,
  ADD_COL_SUCCESS,
  ADD_COL_FAILURE,
} from "../actions/addColumn";
import {
  ADD_BOARD_START,
  ADD_BOARD_SUCCESS,
  ADD_BOARD_FAILURE,
} from "../actions/addBoard";
import {
  FETCH_BOARD_START,
  FETCH_BOARD_SUCCESS,
  FETCH_BOARD_FAILURE,
} from "../actions/fetchBoardByBoardId";
import {
  FETCH_BOARDCOLS_START,
  FETCH_BOARDCOLS_SUCCESS,
  FETCH_BOARDCOLS_FAILURE,
} from "../actions/fetchColumnsByBoardId";
import {
  FETCH_BOARDITEMS_START,
  FETCH_BOARDITEMS_SUCCESS,
  FETCH_BOARDITEMS_FAILURE,
} from "../actions/fetchItemsByBoardId";
import {
  ADD_ITEM_START,
  ADD_ITEM_SUCCESS,
  ADD_ITEM_FAILURE,
} from "../actions/addItem";
import {
  FETCH_MEMBERS_START,
  FETCH_MEMBERS_SUCCESS,
  FETCH_MEMBERS_FAILURE,
} from "../actions/fetchBoardMembersByBoardId";

const experimentalBoardState = {
  //https://stackoverflow.com/questions/34401098/remove-a-property-in-an-object-immutably
  addColumnStart: false,
  addColumnSuccess: false,
  addColumnFailure: "",
  addBoardStart: false,
  addBoardSuccess: false,
  addBoardFailure: "",
  addItemStart: false,
  addItemSuccess: false,
  addItemFailure: "",
  activeBoard: null,
  fetchMembersStart: false,
  fetchMembersSuccess: false,
  fetchMembersFailure: false,

  entities: {
    members: {
      byId: {},
    },
    boards: {
      byId: {},
      allIds: [],
    },
    columns: {
      byId: {},
      allIds: [],
    },

    items: {
      byId: {},

      allIds: [],
    },
    columnBoard: {
      byId: {},
      allIds: [],
    },
    itemColumn: {
      byId: {},
      allIds: [],
    },
  },
};

export const expBoardReducer = (state = experimentalBoardState, action) => {
  const { entities } = state;
  const { boards, columns, items, columnBoard, itemColumn, members } = entities;
  switch (action.type) {
    case ADD_BOARD_START:
      return {
        ...state,
        addBoardStart: true,
        addBoardSuccess: false,
        addBoardFailure: false,
      };
    case ADD_BOARD_SUCCESS:
      return {
        ...state,
        addBoardStart: false,
        addBoardSuccess: true,
        activeBoard: action.payload.board_id,
        entities: {
          ...entities,
          boards: {
            ...boards,
            byId: {
              ...boards.byId,
              [action.payload.board_id]: action.payload,
            },
            allIds: [...boards.allIds, action.payload.board_id],
          },
        },
      };
    case ADD_BOARD_FAILURE:
      return {
        ...state,
        addBoardStart: false,
        addBoardSuccess: false,
        addBoardFailure: action.payload,
      };
    case FETCH_BOARD_START:
      return {
        ...state,
        fetchBoardStart: true,
        fetchBoardSuccess: false,
        fetchBoardFailure: false,
      };
    case FETCH_BOARD_SUCCESS:
      return {
        ...state,
        fetchBoardStart: false,
        fetchBoardSuccess: true,
        activeBoard: action.payload.board_id,
        entities: {
          ...entities,
          boards: {
            ...boards,
            byId: {
              ...boards.byId,
              [action.payload.board_id]: action.payload,
            },
            allIds: [...boards.allIds, action.payload.board_id],
          },
        },
      };
    case FETCH_BOARD_FAILURE:
      return {
        ...state,
        fetchBoardStart: false,
        fetchBoardSuccess: false,
        fetchBoardFailure: action.payload,
      };

    case ADD_COL_START:
      return {
        ...state,
        addColumnStart: true,
        addColumnSuccess: false,
        addColumnFailure: "",
      };
    case ADD_COL_SUCCESS:
      return {
        ...state,
        addColumnStart: false,
        addColumnSuccess: true,
        entities: {
          ...entities,
          columns: {
            ...columns,
            byId: {
              ...columns.byId,
              [action.payload.column_id]: action.payload,
            },
            allIds: [...columns.allIds, action.payload.column_id],
          },
          columnBoard: {
            ...columnBoard,
            byId: {
              ...columnBoard.byId,
              [uuid4()]: action.payload,
            },
          },
        },
      };
    case ADD_COL_FAILURE:
      return {
        ...state,
        addColumnStart: false,
        addColumnSuccess: false,
        addColumnFailure: action.payload,
      };
    case FETCH_BOARDCOLS_START:
      return {
        ...state,
        fetchColumnStart: true,
        fetchColumnSuccess: false,
        fetchColumnFailure: "",
      };
    case FETCH_BOARDCOLS_SUCCESS:
      return {
        ...state,
        fetchColumnStart: false,
        fetchColumnSuccess: true,
        entities: {
          ...entities,
          columns: {
            ...columns,
            byId: {
              ...columns.byId,
              ...action.payload.reduce((obj, v) => {
                obj[v.column_id] = v;
                return obj;
              }, {}),
            },
          },
          columnBoard: {
            ...columnBoard,
            byId: {
              ...columnBoard.byId,
              ...action.payload.reduce((obj, v) => {
                obj[uuid4()] = v;
                return obj;
              }, {}),
            },
          },
          allIds: [...columns.allIds, action.payload.column_id],
        },
      };
    case FETCH_BOARDCOLS_FAILURE:
      return {
        ...state,
        fetchColumnStart: false,
        fetchColumnSuccess: false,
        fetchColumnFailure: action.payload,
      };

    case ADD_ITEM_START:
      return {
        ...state,
        addItemStart: true,
        addItemSuccess: false,
        addItemFailure: "",
      };
    case ADD_ITEM_SUCCESS:
      return {
        ...state,
        addItemStart: false,
        addItemSuccess: true,
        entities: {
          ...entities,
          items: {
            ...items,
            byId: {
              ...items.byId,
              [action.payload.item_id]: action.payload,
            },
            allIds: [...items.allIds, action.payload.item_id],
          },
          itemColumn: {
            ...itemColumn,
            byId: {
              ...itemColumn.byId,
              [action.payload.item_id]: action.payload,
            },
          },
        },
      };
    case ADD_ITEM_FAILURE:
      return {
        ...state,
        addItemStart: false,
        addItemSuccess: false,
        addItemFailure: action.payload,
      };
    case FETCH_BOARDITEMS_START:
      return {
        ...state,
        fetchItemStart: true,
        fetchItemSuccess: false,
        fetchItemFailure: "",
      };
    case FETCH_BOARDITEMS_SUCCESS:
      return {
        ...state,
        fetchItemStart: false,
        fetchItemSuccess: true,
        entities: {
          ...entities,
          items: {
            ...items,
            byId: {
              ...items.byId,
              ...action.payload.reduce((obj, v) => {
                obj[v.item_id] = v;
                return obj;
              }, {}),
            },
          },
          itemColumn: {
            ...itemColumn,
            byId: {
              ...itemColumn.byId,
              ...action.payload.reduce((obj, v) => {
                obj[v.item_id] = v;
                return obj;
              }, {}),
            },
          },
          allIds: [...items.allIds, action.payload.item_id],
        },
      };
    case FETCH_BOARDITEMS_FAILURE:
      return {
        ...state,
        fetchItemStart: false,
        fetchItemSuccess: false,
        fetchItemFailure: action.payload,
      };
    case "RESET":
      return {
        ...experimentalBoardState,
      };
    case "SOCKET_CLOSE":
      members.byId[action.payload.userId]["isActive"] = action.payload.status;

      return {
        ...state,
        entities: {
          ...entities,
          members: {
            ...members,
            byId: {
              ...members.byId,
              [action.payload.userId]: member,
            },
          },
        },
      };

    case "SOCKET_OPEN":
      let member = members.byId[action.payload.userId];
      member["isActive"] = action.payload.status;
      return {
        ...state,
        entities: {
          ...entities,
          members: {
            ...members,
            byId: {
              ...members.byId,
              [action.payload.userId]: member,
            },
          },
        },
      };

    case FETCH_MEMBERS_START:
      return {
        ...state,
        fetchMembersStart: true,
      };
    case FETCH_MEMBERS_SUCCESS:
      return {
        ...state,
        fetchMembersStart: false,
        fetchMembersSuccess: true,
        entities: {
          ...entities,
          members: {
            ...members,
            byId: {
              ...action.payload.reduce((obj, curr) => {
                obj[curr.user_uuid] = curr;
                return obj;
              }, {}),
            },
          },
        },
      };
    case FETCH_MEMBERS_FAILURE:
      return {
        ...state,
        fetchMembersStart: false,
        fetchMembersSuccess: false,
        fetchMembersFailure: true,
      };

    default:
      return state;
  }
};
