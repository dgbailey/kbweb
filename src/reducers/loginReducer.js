
const initialState = {
    loginStart:false,
    loginSuccess:false,
    loginError:null,
    loginPayload:null
}

export const loginReducer = (state=initialState,action) =>{

    switch(action.type){

        case 'PREFLIGHT_START':
            return {
                loginStart:true,
                loginSuccess:false,
                loginError:null,

            }
        case 'PREFLIGHT_SUCCESS':
            return {
                loginStart:false,
                loginSuccess:true,
                loginError:null,
                loginPayload:action.payload
            }

        case 'PREFLIGHT_FAILURE':
            return {
                loginStart:false,
                loginSuccess:false,
                loginPayload:null,
                loginError:action.payload,
            }

        case 'START_login':
            return {
                ...state,
                loginStart:true,
                loginStart:false,
                loginError:null,
                loginPayload:null
            }

        case 'login_COMPLETE':
            return {
                ...state,
                loginStart:false,
                loginSuccess:true,
                loginPayload:action.payload
            }
        
        case 'login_ERROR':
            return {
                ...state,
                loginStart:false,
                loginSuccess:false,
                loginPayload:null,
                loginError:action.payload
            }
        
        default:
            return state
    }



}