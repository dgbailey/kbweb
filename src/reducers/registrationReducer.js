const initialState = {
	registrationStart: false,
	registrationSuccess: false,
	registrationError: ''
};

export const registrationReducer = (state = initialState, action) => {
	switch (action.type) {
		case 'START_REGISTRATION':
			return {
				...state,
				registrationStart: true,
				registrationStart: false,
				registrationError: ''
			};

		case 'REGISTRATION_COMPLETE':
			return {
				...state,
				registrationStart: false,
				registrationSuccess: true
			};

		case 'REGISTRATION_ERROR':
			return {
				...state,
				registrationStart: false,
				registrationSuccess: false,
				registrationError: action.payload
			};

		default:
			return state;
	}
};
