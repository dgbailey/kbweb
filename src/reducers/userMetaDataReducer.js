const initialState = {
	id: null,
	boardIds: [],
	userName: null,
	fetchStart: false,
	fetchSuccess: false,
	fetchFailure: false
};

const FETCHING_START_METADATA = 'FETCHING_USER_METADATA';
const FETCHING_SUCCESS_METADATA = 'FETCHING_SUCCESS_METADATA';
const FETCHING_FAILURE_METADATA = 'FETCHING_USER_METADATA';

export const userMetaDataReducer = (state = initialState, action) => {
	switch (action.type) {
		case FETCHING_START_METADATA:
			return {
				...state,
				fetchStart: true
			};
		case FETCHING_SUCCESS_METADATA:
			return {
				...state,
				fetchStart: false,
				fetchSuccess: true,
				id: action.payload.userId,
				userName: action.payload.userName,
				boardIds: action.payload.metaData || []
			};
		case FETCHING_FAILURE_METADATA:
			return {
				...state,
				fetchStart: false,
				fetchSuccess: true,
				id: null,
				boardIds: [],
				fetchFailure: action.payload
			};
		default:
			return state;
	}
};
