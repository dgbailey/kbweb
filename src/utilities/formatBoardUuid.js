
export const formatBoardUuid = (uuid) => {
    return uuid.replace(/-/g,"");
}


