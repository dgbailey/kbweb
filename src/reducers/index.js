import { combineReducers } from 'redux';
import { boardReducer } from './boardReducer';
import { registrationReducer } from './registrationReducer';
import { loginReducer } from './loginReducer';
import { userMetaDataReducer } from './userMetaDataReducer';
import { expBoardReducer } from './boardReducer';
import { boardMemberReducer } from './boardMembersReducer';
export const rootReducer = combineReducers({
	registrationStatus: registrationReducer,
	loginStatus: loginReducer,
	userMetaData: userMetaDataReducer,
	expBoard: expBoardReducer,
	addMemberStatus: boardMemberReducer
});
