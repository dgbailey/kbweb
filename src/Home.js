import React from 'react';
import { useSelector } from 'react-redux';
import styled from 'styled-components';
import { Preview } from './Preview';
import { formatBoardUuid } from './utilities/formatBoardUuid';
import {addBoard} from './actions/addBoard';
import {useHistory} from 'react-router-dom';
import {useDispatch} from 'react-redux';

export const Home = () => {

	const boardMetaData = useSelector((state) => state.userMetaData.boardIds);
	const dispatch = useDispatch();
	const history = useHistory();
	const boardURIprefix = '/board/existing';
	const userMetaData = useSelector((state) => state.userMetaData);
	const { id: userId } = userMetaData;
	const name = 'Get-Started';

	async function addNewBoard(){
		let board = await addBoard({userId,name},dispatch);
		let newUri = boardURIprefix + "/" + name + "-" + formatBoardUuid(board.board_id)
		history.push(newUri)

	}
	function renderPreviews(){
		
			return (
				<>
			
				{boardMetaData.map((b) => (
					<Preview
						to={boardURIprefix +'/' + b.name + '-' + formatBoardUuid(b.board_id)}
						key={b.board_id}
						id={b.board_id}
						name={b.name}
					/>
				))}
				</>
			)
		}
	

	return (
		<StyledHome>
			<button onClick={addNewBoard}>Create New Board</button>
			{boardMetaData.length > 0 && renderPreviews()}
		</StyledHome>
	)
};

const StyledHome = styled.section`
	height: 100%;
	width: 100%;
	border: 1px solid red;
	display: flex;
	flex-wrap: wrap;
`;
