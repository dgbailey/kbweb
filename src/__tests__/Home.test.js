import React from 'react';
import { TestingContext } from '../utilities/testing/configContext';
import { Home } from '../Home';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import { initialState } from '../reducers/userMetaDataReducer/';
import uuid4 from 'uuid4';
import * as formatUuid from '../utilities/formatBoardUuid';
const formattingSpy = jest.spyOn(formatUuid, 'formatBoardUuid');

describe('Home component Tests', () => {
	let boardIds = [
		' 87242b88-d14a-4dfb-a377-b799a619880e',
		'87242b88-d14a-4dfb-a377-b799a639880e',
		'87242b88-d14a-4dfb-a377-b792a619880e '
	].map((id) => ({ board_id: id, name: 'generic' }));
	let wrapper;
	let context;
	beforeAll(() => {
		context = (
			<TestingContext storeState={{ userMetaData: { ...initialState, boardIds } }}>
				<Home />
			</TestingContext>
		);
		wrapper = mount(context);
	});

	test('Tests mount', () => {
		expect(wrapper.containsMatchingElement(<Home />)).toBe(true);
	});

	test(`Tests Uuid utility to be called ${boardIds.length} times.`, () => {
		expect(formattingSpy).toHaveBeenCalledTimes(boardIds.length);
	});

	test('Home correctly with appropriate ${boardIds.length} Preview Children', () => {
		//uuids will change snapshot every test
		const tree = renderer.create(wrapper).toJSON();
		expect(tree).toMatchSnapshot();
	});

	// test('Tests Create new board function called on button click', () => {
	// 	let createBoardSpy = jest.spyOn(Home, 'addNewBoard');
	// 	wrapper.find('button').simulate('click');
	// 	expect(createBoardSpy).toHaveBeenCalledTimes(1);
	// });

	//Is it necessary to test functions within functional components? Is it possible?
});
