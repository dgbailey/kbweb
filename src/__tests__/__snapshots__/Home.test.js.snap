// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home component Tests Home correctly with appropriate \${boardIds.length} Preview Children 1`] = `
<section
  className="sc-bwzfXH jIBrOh"
>
  <button
    onClick={[Function]}
  >
    Create New Board
  </button>
  <a
    href="/board/existing/generic- 87242b88d14a4dfba377b799a619880e"
    onClick={[Function]}
  >
    <ul
      className="sc-bdVaJa gAYVjo"
    >
      <li>
        <h2
          className="name"
        >
          generic
        </h2>
      </li>
    </ul>
  </a>
  <a
    href="/board/existing/generic-87242b88d14a4dfba377b799a639880e"
    onClick={[Function]}
  >
    <ul
      className="sc-bdVaJa gAYVjo"
    >
      <li>
        <h2
          className="name"
        >
          generic
        </h2>
      </li>
    </ul>
  </a>
  <a
    href="/board/existing/generic-87242b88d14a4dfba377b792a619880e "
    onClick={[Function]}
  >
    <ul
      className="sc-bdVaJa gAYVjo"
    >
      <li>
        <h2
          className="name"
        >
          generic
        </h2>
      </li>
    </ul>
  </a>
</section>
`;
